File: ios/HybridVisionCameraPluginInatVision.swift
import AVFoundation
import NitroModules
import VisionCamera

class HybridVisionCameraPluginInatVision: HybridVisionCameraPluginInatVisionSpec {
  private let plugin = VisionCameraPluginInatVisionPlugin()

  func call(frame: any HybridFrameSpec, options: VisionCameraPluginInatVisionOptions) throws -> AnyMap {

    let sampleBuffer = try sampleBuffer(from: frame)
    let pixelBuffer = try pixelBuffer(from: sampleBuffer)

    let args = options.toDictionary()
    let response = plugin.callback(
      pixelBuffer,
      withArguments: args as [String: Any]
    )
    let dict = jsonSafe(response) as? [String: Any] ?? [:]
    return AnyMap.fromDictionaryIgnoreIncompatible(
      dict.mapValues { $0 as Any? }
    )
  }

  private func jsonSafe(_ value: Any?) -> Any? {
    switch value {
    case nil, is NSNull:
      return nil
    case let dict as NSDictionary:
      var out: [String: Any] = [:]
      for (key, nested) in dict {
        if let key = key as? String, let safe = jsonSafe(nested) {
          out[key] = safe
        }
      }
      return out
    case let arr as NSArray:
      return arr.compactMap { jsonSafe($0) }
    default:
      return value
    }
  }

  private func sampleBuffer(from frame: any HybridFrameSpec) throws -> CMSampleBuffer {
    guard let nativeFrame = frame as? any NativeFrame else {
      throw RuntimeError.error(withMessage: "The given Frame is not of type `NativeFrame`!")
    }
    guard let sampleBuffer = nativeFrame.sampleBuffer else {
      throw RuntimeError.error(withMessage: "The given Frame's `sampleBuffer` is no longer valid!")
    }
    return sampleBuffer
  }

  private func pixelBuffer(from sampleBuffer: CMSampleBuffer) throws -> CVPixelBuffer {
    guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
      throw RuntimeError.error(
        withMessage: "The given Frame does not contain a valid image buffer!")
    }
    return pixelBuffer
  }
}

File: android/src/main/java/com/margelo/nitro/visioncameraplugininatvision/HybridVisionCameraPluginInatVision.kt
package com.margelo.nitro.visioncameraplugininatvision

// import android.media.Image
// import androidx.camera.core.ExperimentalGetImage
import com.margelo.nitro.camera.HybridFrameSpec
// import com.margelo.nitro.camera.public.NativeFrame
import com.margelo.nitro.core.AnyMap

// import com.visioncameraplugininatvision.VisionCameraPluginInatVisionPlugin

class HybridVisionCameraPluginInatVision : HybridVisionCameraPluginInatVisionSpec() {
  // private val plugin = VisionCameraPluginInatVisionPlugin()

  override fun call(frame: HybridFrameSpec, options: AnyMap): AnyMap {
    // val image = image(from = frame)
    // val response = plugin.callback(image, options.toHashMap())
    // val dict = jsonSafe(response) as? Map<String, Any> ?: emptyMap()
    // return AnyMap.fromMap(dict, true)
    return options;
  }

//  private fun image(from: HybridFrameSpec): Image {
//    val nativeFrame =
//      from as? NativeFrame
//        ?: throw Error("The given Frame is not of type `NativeFrame`!")
//    val image =
//      nativeFrame.image.image
//        ?: throw Error("The given Frame's `image` is no longer valid!")
//    return image
//  }

//  private fun jsonSafe(value: Any?): Any? {
//    if (value == null) {
//      return null
//    }
//    return when (value) {
//      is Map<*, *> -> {
//        val out = LinkedHashMap<String, Any>()
//        for ((key, nested) in value) {
//          if (key is String) {
//            jsonSafe(nested)?.let { out[key] = it }
//          }
//        }
//        out
//      }
//      is List<*> -> value.mapNotNull { jsonSafe(it) }
//      else -> value
//    }
//  }
}
